### PR TITLE
Add SpanSet set operations with Masks

### PR DESCRIPTION
**Pull request by pschella**
Introduce method overloads which enable set operations between a
SpanSet and a Mask. Additionally add a free function to turn a
Mask into a SpanSet.

### CODE REVIEW

File: include/lsst/afw/geom/SpanSet.h [hunk 1]
```diff
@@ -38,6 +38,21 @@
...
 #include "lsst/afw/geom/SpanSetFunctorGetters.h"
 
+namespace lsst { namespace afw { namespace geom { namespace details {
+    /* Functor object to be used with maskToSpanSet function
+     */
+    template <typename T>
+    class MaskToSpanSetDefaultFunctor {
+     public:
+        bool operator()(T const & pixelValue) {
+            return pixelValue ? true : false;
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: I would prefer `return pixelValue != 0` but it doesn't really matter much.

File: include/lsst/afw/geom/SpanSet.h [hunk 2]
```diff
@@ -38,6 +38,21 @@
 #include "lsst/afw/geom/Box.h"
 #include "lsst/afw/geom/SpanSetFunctorGetters.h"
 
+namespace lsst { namespace afw { namespace geom { namespace details {
+    /* Functor object to be used with maskToSpanSet function
+     */
+    template <typename T>
+    class MaskToSpanSetDefaultFunctor {
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: I would give this a more descriptive name such as `AnyBitSet`.

File: include/lsst/afw/geom/SpanSet.h [hunk 3]
```diff
@@ -48,6 +63,63 @@ namespace lsst { namespace afw { namespace geom {
...
+ * @param comparator - Functor object to use in the decision to include pixel in SpanSet. Should return
+ *                     true when a given pixel in the mask should be part of the SpanSet, and false
+ *                     otherwise. The functor takes a single value taken from the mask at the
+ *                     pixel under consideration. Defaults to evaluating true if the mask has bits set,
+ *                     and false otherwise.
+ *
+ * @tparam T - Pixel type of the Mask
+ * @tparam F - Type of the functor
+ */
+template <typename T, typename F = details::MaskToSpanSetDefaultFunctor<T>>
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: Rename `F` to `Predicate`.

File: include/lsst/afw/geom/SpanSet.h [hunk 4]
```diff
@@ -48,6 +63,63 @@ namespace lsst { namespace afw { namespace geom {
...
+std::shared_ptr<geom::SpanSet> maskToSpanSet(image::Mask<T> const & mask,
+                               F comparator=details::MaskToSpanSetDefaultFunctor<T>()) {
+    std::vector<Span> tempVec;
+    std::size_t startValue{0};
+    bool started{false};
+    auto const & maskArray = mask.getArray();
+    auto dimensions = maskArray.getShape();
+    for (size_t y{0}; y < dimensions[0]; ++y) {
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: Fine with me, but I expect most people to want to see `size_t y = 0` here. Alternatively you could consider `auto y = size_t{0}`.

File: include/lsst/afw/geom/SpanSet.h [hunk 5]
```diff
@@ -48,6 +63,63 @@ namespace lsst { namespace afw { namespace geom {
...
+        startValue = 0;
+        started = false;
+        for (size_t x{0}; x < dimensions[1]; ++x) {
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: Same here.

File: include/lsst/afw/geom/SpanSet.h [hunk 6]
```diff
@@ -48,6 +63,63 @@ namespace lsst { namespace afw { namespace geom {
...
+            // If a new span has not been started, and a given x matches the functor condition
+            // start a new span
+            if (comparator(maskArray[y][x]) && !started) {
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: Switching conditions around may result in fewer tests.

File: include/lsst/afw/geom/SpanSet.h [hunk 7]
```diff
@@ -48,6 +63,63 @@ namespace lsst { namespace afw { namespace geom {
...
+            // Span being created should be stopped, and appended to the Span vector
+            else if (!comparator(maskArray[y][x]) && started) {
+                tempVec.push_back(Span(y, startValue, x-1));
+                started = false;
+            }
+            // If this is the last value in the Span's x range, and started is still true
+            // that means the last value does not evaluate false in the functor and should be
+            // included in the Span under construction. The Span should be completed and added
+            // to the Span Vector before the next span is concidered.
+            if (x == dimensions[1] && started) {
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: Switching conditions around may result in fewer tests.

File: include/lsst/afw/geom/SpanSet.h [hunk 8]
```diff
@@ -457,19 +529,69 @@ class SpanSet {
...
+     * @param other - Mask with which to calculate instersection
+     * @param bitmask - The bit value to concider when intersecting
+     *
+     * @tparam T - Pixel type of the Mask
+     */
+    template <typename T>
+    std::shared_ptr<SpanSet> intersectNot(image::Mask<T> const & other, T const & bitmask) const {
+        auto comparator = [bitmask]
+                          (T pixelValue)
+                          {return (pixelValue & bitmask) == bitmask ? true : false;};
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: Same here.

File: include/lsst/afw/geom/SpanSet.h [hunk 9]
```diff
@@ -457,19 +529,69 @@ class SpanSet {
...
+     * @param other - Mask with which to calculate instersection
+     * @param bitmask - The bit value to concider when intersecting
+     *
+     * @tparam T - Pixel type of the Mask
+     */
+    template <typename T>
+    std::shared_ptr<SpanSet> union_(image::Mask<T> const & other, T const & bitmask) const {
+        auto comparator = [bitmask]
+                          (T pixelValue)
+                          {return (pixelValue & bitmask) == bitmask ? true : false;};
```
Comment (pschella) [inline include/lsst/afw/geom/SpanSet.h]: Here.

File: tests/testSpanSets.cc [hunk 10]
```diff
@@ -477,6 +477,25 @@ std::pair<std::shared_ptr<afwGeom::SpanSet>, std::shared_ptr<afwGeom::SpanSet>>
                      std::shared_ptr<afwGeom::SpanSet>>(firstSpanSet, secondSpanSet);
 }
 
+std::pair<lsst::afw::image::Mask<lsst::afw::image::MaskPixel>, std::shared_ptr<afwGeom::SpanSet> >
```
Comment (pschella) [inline tests/testSpanSets.cc]: No need for extra space in `> >`.

File: tests/testSpanSets.cc [hunk 11]
```diff
@@ -477,6 +477,25 @@ std::pair<std::shared_ptr<afwGeom::SpanSet>, std::shared_ptr<afwGeom::SpanSet>>
...
+ makeMaskAndSpanSetForOperationTests() {
+    // Create three overlapping regions in a mask, and a SpanSet to use in set operation tests
+    // This box will range from 0 to 4 in y and 0 to 4 in x
```
Comment (pschella) [inline tests/testSpanSets.cc]: 0 to 4? I'm confused. Shouldn't it be 2 to 4?

File: include/lsst/afw/geom/SpanSet.h (hunk 4 above)
Comment (TallJimbo) [inline include/lsst/afw/geom/SpanSet.h]: See also [RFC-242](https://jira.lsstcorp.org/browse/RFC-242) (which is unresolved, but I think this case is an argument in favor of my proposal to prefer C++98 syntax for initialization).

File: tests/testSpanSets.cc (hunk 11 above)
Comment (natelust) [inline tests/testSpanSets.cc]: the first argument in From Shape is a radius. So spanSetFromShape(2, BOX) will make a box that ranges from -2 to 2 in both x and y. Shifting by 2,2 will bring the lower corner up to 0,0 and the upper corner to 4,4.

File: tests/testSpanSets.cc (hunk 10 above)
Comment (natelust) [inline tests/testSpanSets.cc]: changed

File: include/lsst/afw/geom/SpanSet.h (hunk 1 above)
Comment (natelust) [inline include/lsst/afw/geom/SpanSet.h]: changed

File: include/lsst/afw/geom/SpanSet.h (hunk 2 above)
Comment (natelust) [inline include/lsst/afw/geom/SpanSet.h]: changed to AnyBitSetFunctor

File: include/lsst/afw/geom/SpanSet.h (hunk 3 above)
Comment (natelust) [inline include/lsst/afw/geom/SpanSet.h]: done

File: include/lsst/afw/geom/SpanSet.h [hunk 12]
```diff
@@ -48,6 +63,63 @@ namespace lsst { namespace afw { namespace geom {
...
+ *                     otherwise. The functor takes a single value taken from the mask at the
+ *                     pixel under consideration. Defaults to evaluating true if the mask has bits set,
+ *                     and false otherwise.
+ *
+ * @tparam T - Pixel type of the Mask
+ * @tparam F - Type of the functor
+ */
+template <typename T, typename F = details::MaskToSpanSetDefaultFunctor<T>>
+std::shared_ptr<geom::SpanSet> maskToSpanSet(image::Mask<T> const & mask,
+                               F comparator=details::MaskToSpanSetDefaultFunctor<T>()) {
```
Comment (natelust) [inline include/lsst/afw/geom/SpanSet.h]: done

**Pull request merged**